### PR TITLE
Add Rubi Numpad

### PR DESCRIPTION
**Pull request by ohchiko**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/rubi/lib/glcdfont.c [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+#pragma once
+
+#ifdef __AVR__
+ #include <avr/io.h>
+ #include <avr/pgmspace.h>
+#elif defined(ESP8266)
+ #include <pgmspace.h>
+#else
+ #define PROGMEM
+#endif
```
Comment (drashna) [inline keyboards/rubi/lib/glcdfont.c:27]: You can replace all of this with just: 
```suggestion
#include "progmem.h"
```

File: keyboards/rubi/lib/glcdfont.c [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  0x77, 0x77, 0x3E, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x3E, 0x6B, 0x77, 0x41,
+  0x77, 0x6B, 0x3E, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x4C, 0x52, 0x52, 0x52,
+  0x24, 0x00, 0x3C, 0x42, 0x42, 0x42,
+  0x24, 0x00, 0x7E, 0x12, 0x32, 0x52,
+  0x4C, 0x00, 0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00, 0x26, 0x10,
+  0x08, 0x04, 0x00, 0x00, 0x00, 0x00,
+};
```
Comment (drashna) [inline keyboards/rubi/lib/glcdfont.c:255]: if you need more characters, you can get 22 more.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/rubi/lib/glcdfont.c (hunk 2 above)
Comment (ohchiko) [reply keyboards/rubi/lib/glcdfont.c:246]: I think this is enough. Actually I was just copied from another board, then modified it to fit my needs.

**Review by drashna — APPROVED** [review]

**Comment by ohchiko** [comment]
Any other changes needed for this PR?

File: keyboards/rubi/rubi.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    current_layer = get_highest_layer(state);
+    return state;
+}
+
+bool led_update_kb(led_t led_state) {
+    writePin(C6, led_state.num_lock);
+    return true;
+}
+
+void encoder_update_kb(uint8_t index, bool clockwise) {
```
Comment (drashna) [inline keyboards/rubi/rubi.c:79]: No user function?

File: keyboards/rubi/rubi.c [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "rubi.h"
+
```
Comment (drashna) [inline keyboards/rubi/rubi.c:17]: Most of the functions here don't have the user function called, meaning that users cannot control the functionality. 

I would prefer that that the user functions be used here, and set to weak, so they can be overridden.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/rubi/rubi.c (hunk 4 above)
Comment (ohchiko) [reply keyboards/rubi/rubi.c:17]: Ah, I forgot to call the user functions. I could just call them before or after the keyboard function, right? For example, I could set the `state` variable for `layer_state_set_kb` function to the returned value of `layer_state_set_user` function, like below:
```
layer_state_t layer_state_set_kb(layer_staate_t state) {
    state = layer_state_set_user(state);
    current_layer = get_highest_layer(state);
    return state;
}
```

Comment (ohchiko) [reply keyboards/rubi/rubi.c:17]: I choose the same way for the `led_update_kb` function. For the `encoder_update_kb` function, as your suggestion I moved its content to a weak `encoder_update_user` function and call it in the `encoder_update_user` function. Any other things I need to do before I push this changes?

Comment (drashna) [reply keyboards/rubi/rubi.c:17]: That's the best way, for the most part.  

Though, `current_layer` isn't needed.  `layer_state` is a globally available bitmask of the layer state, so it is accessible anywhere.

Comment (ohchiko) [reply keyboards/rubi/rubi.c:17]: Implemented in https://github.com/qmk/qmk_firmware/pull/12283/commits/9cf41ee489b38ce8e1e07a85ac57f20674a23488.

File: keyboards/rubi/lib/oled.c [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+                oled_off();
+            }
+        }
+    } else {
+        render_logo();
+        oled_logo_expired = timer_elapsed(oled_logo_timer) > OLED_LOGO_TIMEOUT;
+    }
+}
+
+void oled_task_user(void) {
```
Comment (drashna) [inline keyboards/rubi/lib/oled.c:242]:
```suggestion
__attribute__((weak)) void oled_task_user(void) {
```

File: keyboards/rubi/rubi.c [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        default:
+            break;
+    }
+
+    return process_record_user_oled(keycode, record);
+}
+
+layer_state_t layer_state_set_kb(layer_state_t state) {
+    return layer_state_set_user(state);
+}
```
Comment (drashna) [inline keyboards/rubi/rubi.c:69]:
```suggestion
```

File: keyboards/rubi/rubi.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+#include "quantum.h"
+
+#define CALC_DIGITS 12
+
+enum rubi_keycodes {
+    ENC_PRESS = SAFE_RANGE,
+    CL_PLUS,
+    CL_STAR,
+    CL_TYPE
```
Comment (drashna) [inline keyboards/rubi/rubi.h:26]:
```suggestion
    CL_TYPE,
    NEW_SAFE_RANGE
```

File: keyboards/rubi/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+// Defines names for use in layer keycodes and the keymap
```
Comment (drashna) [inline keyboards/rubi/keymaps/default/keymap.c:18]:
```suggestion

// enum custom_keycodes {
//     FIRST = NEW_SAFE_RANGE,
// };

// Defines names for use in layer keycodes and the keymap
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Comment by ohchiko** [comment]
Does this PR needs any other changes?

File: keyboards/rubi/rules.mk [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
```
Comment (fauxpark) [inline keyboards/rubi/rules.mk:22]:
```suggestion
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/rubi/readme.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+# Rubi
+
+![Rubi](https://i.imgur.com/XUzxe0r.jpg)
```
Comment (fauxpark) [inline keyboards/rubi/readme.md:3]:
```suggestion
![Rubi](https://i.imgur.com/XUzxe0rh.jpg)
```

File: keyboards/rubi/readme.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+![Proto Album](https://imgur.com/a/QVtU7hV)
```
Comment (fauxpark) [inline keyboards/rubi/readme.md:5]:
```suggestion
[Proto Album](https://imgur.com/a/QVtU7hV)
```

File: keyboards/rubi/readme.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    make rubi:default
+
+Flashing example for this keyboard:
+
+    make rubi:default:flash
+
+To put this keyboard into bootloader mode (for flashing firmware), simply activate layer 3 and press the middle key (key "5" on the proto picture) on the default keymap.
+
+![Default Keymap](keymaps/default/readme.md)
+
```
Comment (fauxpark) [inline keyboards/rubi/readme.md:24]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**